### DISCUSSION TITLE
Better integration of Java modules as dependencies in the future of Maven

### DISCUSSION THREAD

**Discussion by hendrikebbers** [support-and-care/maven-support-and-care#38] [Ideas for Funding]
This is based on the discussion in https://github.com/OpenElements/maven-support-care/discussions/11

We need to discuss how the Java module system can have a better integration in the Maven dependency definitions. Especially since maven dependecies and java modules have some common definitions but next to that some definitions that might nit work perfectlly well together.

Goal of this milestone should not be to have a final 100% solid integration in Maven but to have a definition for the future with a working prototype. Next to that parts of the prototype that does not make Maven 4 incompatible can be merged directly to Maven.

**Comment by desruisseaux** [comment]
There is one functionality that we may need to add in Maven core: the capability to get the JPMS name of a JAR dependency. It involves opening the JAR file and reading its module descriptor. Because many plugins may need this information, handling that in Maven core would allow to cache the JPMS module name after the first request.

**Reply by rmannibucau** [reply]
> There is one functionality that we may need to add in Maven core

-> There is one functionality that we may need to add in Maven *shared tools* ? core don't care at all about all that stuff and it is not its responsability.
Ultimately it could be a maven plugin api service but my 2 cts is that it is very localised for a few tools (toolprovider) and not that generic so would fit very well a plugin shared lib too.
Does not prevent the caching in the maven session - agree it would be good to cache it.

**Comment by jjohannes** [comment]
Here is a breakdown of what (I think) should be done here.

Create a [Maven Extension](https://maven.apache.org/guides/mini/guide-using-extensions.html) that adopts features from the similar [Gradle plugin](https://github.com/gradlex-org/java-module-dependencies) to Maven with the following features:

- (1) Building a multi-project (without publishing):
  - Parse the `module-info.java` file where present
  - Map each `requires` to a dependency
    - For Modules in other Maven submodules in the same build, the information should be determined from the project model
    - For 3d party dependency, the [mapping properties file](https://github.com/gradlex-org/java-module-dependencies/blob/main/src/main/resources/org/gradlex/javamodule/dependencies/unique_modules.properties) from the Gradle plugin can be reused
    - There should be an option to place additional mappings in a property file in the project 
    - The scope should be mapped as good as possible (`requires transitive` -> `compile`, `requires static` -> `provided`)
  - Versions need to be added to 3rd party the dependencies (from `<dependencyManagement>` block of POM or parent POM or imported BOM). Ideally, the mechanism that merges the versions into dependencies defined in a POM can be reused. 
- (2) Publishing a project
  - The dependencies should end up in the consumer POM, ideally the scope can be different in some cases: `requires` maps to `compile` in project POM, but to `runtime` in consumer POM
- (3) Dealing with conflicting modules (several modules with the same name)
  - List different kinds of conflicts that are typical in Java and document how to deal with them in Maven using existing mechanics such as _exclude_ and _profiles_ (to select between alternatives)
  - Write proposals for possible enhancements in `<dependencyManagement>` to improve detection and resolution of certain kinds of conflicts
- (4) Dealing with incomplete Jars
  - Show how Jars can be patched if `module-info.class` is missing. Use what is available in the [ModiTect plugin](https://github.com/moditect/moditect).(https://github.com/moditect/moditect).
  - Write proposal of possible enhancements to the Jar patching mechanism
- (5) Testing
  - Show how Whtiebox and Balckbox testing is done. Use what is available in the [ModiTect plugin](https://github.com/moditect/moditect).
  - Write proposal of possible enhancements for testing Java Modules.
- (6) Write a guide how to setup a Java Module System centered project with Maven.

In the end, this will show how a small, typical Java Module System centered project like https://github.com/gradlex-org/java-module-dependencies would look in this setup and how other use cases we have in a real project like https://github.com/hashgraph/hedera-services are solved (and provide a list of things that cannot be easily solved, if any).

**Comment by desruisseaux** [comment]
Can we split the work in two phases? First, allow JPMS support in projects having all the required `module-info` with no conflict. Then in a second phase, develop plugins for managing missing `module-info` and conflicts.

**Reply by rmannibucau** [reply]
While all the data stay in the pom and there is never any need to read both the moduleinfo and the consumer pom to get a valid model it sounds saner to split.

My 2cts would be that jpms is already supported (recall @sormuras explaining it) even if default is not always friendly but switching globally (does not make sense locally) the default is likely an issue for more people than the number it will help since, once again, it just works today.

Side note : 1 is promised to split the model and be bad for users, please dont do.

**Reply by cowwoc** [reply]
@rmannibucau What do you mean by split model? Do you mean that you prefer keeping all the information in `pom.xml` and ignoring the contents of `module-info.java` altogether?

**Reply by sormuras** [reply]
Mind that you (build tool) never generate missing module declarations for me: https://www.sitepoint.com/maven-cannot-generate-module-declaration/

**Reply by rmannibucau** [reply]
@cowwoc in a project it is mainly "your" (ie dev) choice to modelize the model as you want (with the bad IDE integration risk you accept to be very concrete) but on the consumer side the pom will stay the single source of data whatever you do by design. What we absolutely want to avoid is to load the jar to have the model, the consumer pom must stay the absolute single source (+ transitive pom-only).
Now in the project itself - where maven polyglot support is one way to impl module-info support - you are more free but as mentionned 1. justifies to 100% use the pom and then dispatch since module-info is not a dependency management but just a linking descriptor.

@desruisseaux just (almost) dropped the bad heuristic we had between classpath and module path, I don't think it is welcome to add yet another heuristic to map module name to jars (which would be as bad, take spec jars as an example or bundles vs granular deps).

A worse option would be a file mapping since you have the jar in the pom, if it is or not on the module path and the module path in the module-info and it is 100% sufficient, adding a file to do the mapping would create yet another maintenance cost for maven users and yet another source of ambiguity and error.

**Comment by desruisseaux** [comment]
@jjohannes is the property file really necessary? It seems to me that this information can be obtained automatically by reading the module descriptors of each dependency.

@rmannibucau : I propose the opposite of what you said: the ultimate source of true are the `module-info.class` files. The `pom.xml` files would be only complementary (version numbers and URL where to download). It would imply parsing the module descriptors of all dependencies listed in the `<dependencyManagement>` section. But I think that with this parsing, no property file will be needed.

**Reply by cowwoc** [reply]
Instead of downloading the full jar file to look up module-info.java, I assume that clients would download an index/cache of sort from maven central?

**Reply by desruisseaux** [reply]
The JAR would be read from the `.m2` local repository. If a dependency is declared in the `<dependencyManagement>` section, I presume that it is intended to be used. It would be downloaded when first needed and cached in `.m2` as Maven does today for any dependency, than the module descriptors would be read from there.

**Reply by rmannibucau** [reply]
No, the jar would *not* be read since it will be in the pom.
This is the basis of the whole maven ecosystem, including for not jars - please keep in mind maven is not about java by itself.
So the consumer pom will stay for sure the only source of data.
Now, for the community, I hope it also stays the source for the project itself until we intend to not load the project if the project does not compile (like gradle), I'm more than -1 on that.

**Reply by desruisseaux** [reply]
We should clarify if we are talking about the POM used for building the project, or the POM used for describing the artifact deployed in a repository. In Maven 3, this is the same POM. But it has been proposed on the mailing list (not by me) to decouple those two POM files.

For the source code of a modular project, the POM said what each Java module _may_ use (in the `<dependencyManagement>` block), and the `module-info.java` file said what the module _really_ uses (it replaces the `<dependencies>` block, except for tests and optional dependencies). The proposal is that way because if a dependency is not in `module-info` (directly or indirectly), it will be ignored by the JVM no matter what the POM said (unless we force its use with `--add-reads` options).

For the artifact deployed in a repository, the POM would have a `<dependencies>` section generated by the combination of above-cited `<dependencyManagement>` block and `module-info` file. So that POM indeed would contain all the information that Maven needs in order to know which dependencies to download, same as Maven 3 today. Contrarily to the source-code POM (which may use a model version more recent than 4.0.0), the deployed POM would stay a strict 4.0.0 POM model. Note however that in case of disagreement between the POM content and the `module-info` content, again the JVM obey only to the latter. This is why it is proposed as the source of truth.

**Reply by rmannibucau** [reply]
in previous comment "consumer" referenced the consumer pom (deployed) and "local"/"project" one the one in the project.

both have enough big cons to use module-info as source for me to not do it, plus there are still challenges with using module-info as a source (data conflict mentionned earlier - please not yet another resolution strategy).

side note: the consumer pom is done since a few versions.

**Reply by cowwoc** [reply]
Maven might not be only for Java, but it's the happy path. The happy path should be as frictionless as possible while still supporting advanced use-cases.

How about treating `pom.xml` as the source of truth, but... if `module-info.java` **is** present, then pull the values from there. If a property is set both in the `pom.xml` and `module-info.java` the former would override the latter.

With respect to resolving module conflicts, I don't see why we need to do this. If a user's configuration has the same module name from two different transitive dependencies then report their names and fail the build. This is precisely what Java does at load-time when it finds such a conflict.

**Reply by rmannibucau** [reply]
If there is no issue there will be no issue, now we already cases which will fail as mentionned.

Loading a compiled/compilable code is, for now, not an option by default.

All proposed solution are ok and doable in an extension but I dont see them hitting maven before a rebirth from scratch (likely a new project technically) cause user impacts are too high.

Lastly, Im not sure why it is the happy path since not the oracle recommended way...

My 2 cts will be that you can use mx if you want to do all the oracle way - still no idea which feature you are after since at the end you build your app for all cases frictionless today already - once type enhancement integration  will be finished but it is planned.

**Reply by desruisseaux** [reply]
> both have enough big cons to use module-info as source

@rmannibucau , this is not a choice. You can put anything you want in the POM file, no matter if the dependencies appear in the class-path or in the module-path, if a dependency is not in the `module-info` file (potentially indirectly), it will ignored by the JVM, period (unless there is `--add-reads` options). Even the class-path will not work, unless there is an `--add-reads <module name>=ALL-UNNAMED` option. Try it by yourself.

**Reply by rmannibucau** [reply]
This is not related to this topic, just configure the plugins as needed, but this is a runtime concern so a plugin conf by design and as shown, depends of the plugin, not the project.

**Reply by desruisseaux** [reply]
> this is a runtime concern 

No, this is a compile-time **and** a runtime issue. If your dependency is not in `module-info` and there is no `--add-reads` options, your code will not compile. This is not a plugin configuration issue. This is very core to how JPMS work.

**Reply by rmannibucau** [reply]
...the runtime for maven is its plugins so it is a runtime concern so a plugin concern as written...

**Comment by jjohannes** [comment]
FTR: I releases a Maven Extension based on the experiments I did in the context of this discussion last year and partly based on the code base of the corresponding Gradle plugin:

https://github.com/gradlex-org/java-module-dependencies-maven-extension

The extension allows you to omit all `compile` and `provided` scoped dependencies from the `<dependencies>` section of the `pom.xml` of each module in a multi-project. Instead, these will be derived from the `requires` directives in the `module-info.java`.

The extension can be used with Maven 3 or Maven 4. Until now, it is only suitable for application projects that do not publish as the created dependencies are not written back to the POM. I think with Maven 4 though, it should be possible to use the new [ModelTransformer API](https://issues.apache.org/jira/browse/MNG-8232) to change the implementation to create dependency entries that will end up in the consumer POM.
https://github.com/gradlex-org/java-module-dependencies-maven-extension/issues/1

Another interesting topic to explore would be if `pom.xml` files could be omitted if no further individual build configuration is needed for a module. I believe this may be possible to implement in Maven 4 through with the [alternate POM syntaxes support](https://maven.apache.org/whatsnewinmaven4.html#Alternate_POM_syntaxes) but I Have not fully graps that part of Maven's API yet.
https://github.com/gradlex-org/java-module-dependencies-maven-extension/issues/2

**Reply by desruisseaux** [reply]
Thanks for the link. Indeed, deriving the dependencies from the `requires` directives is also something that I plan to do. This is needed for generating the `pom.xml` of the JAR files, exactly the GradleX's issue 1 that you described. We are in agreement. The only nuance is that I plan to restrict this work to the `maven-jar-plugin` at first. Making this automatic derivation more widely visible in the Maven environment could be considered, but it would be another potentially controversial debate. Since I think that this issue is a quasi-blocker only for `maven-jar-plugin`, and since there is already other debates, I would favour the smallest work scope at first.
